kolibri/auth/test/test_permissions.py

```diff
@@ -1,454 +1,568 @@
-#         self.assertFalse(self.coach1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[1]))
-
-#     def test_remove_learner_specific_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[0]))
-
-#     def test_remove_learner_rejects_non_classroom_objects(self):
-#         with self.assertRaises(InvalidPermission):
-#             self.admin.has_perm(self.AUTH_REMOVE_LEARNER, obj={})
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.add_facility_admin     ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_ADD_FACILITY_ADMIN = 'auth.add_facility_admin'
-
-#     def test_add_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.remove_facility_admin  ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_REMOVE_FACILITY_ADMIN = 'auth.remove_facility_admin'
-
-#     def test_remove_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
+"""
+Tests of the permissions on specific models in the auth app. For tests of the permissions system itself, see test_permission_classes.py
+"""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from django.test import TestCase
+
+from .helpers import create_dummy_facility_data
+
+from ..constants import role_kinds
+from ..models import DeviceOwner, Facility, Classroom, LearnerGroup, Role, FacilityUser
+
+
+class FacilityPermissionsTestCase(TestCase):
+    """
+    Tests of permissions for reading/modifying Facility instances
+    """
+
+    def setUp(self):
+        self.data1 = create_dummy_facility_data()
+        self.data2 = create_dummy_facility_data()
+        self.device_owner = DeviceOwner.objects.create(username="boss")
+
+    def test_facility_users_cannot_create_facility(self):
+        """ FacilityUsers can't create new Facilities, regardless of their roles """
+        unsaved_facility = Facility(name="Home")
+        self.assertFalse(unsaved_facility.user_can_create(self.data1["facilityadmin"]))
```

Comment:
> I believe it is standard practice (e.g. in Django Rest Framework) when checking row-level permissions

That makes sense as you can't check row-level permissions if you don't have a row -- but create here is not a row-level permission, right? In contrast to update or read, it applies only to whole Model classes and not rows, if I'm getting it correctly. You could make it a class method, or to go off what I suggest [below](https://github.com/learningequality/kolibri/pull/51#discussion_r58035767) you could pass in the Model class. Incidentally do you have an example of row-level permissions with DRF you can link to, for my own edification?

> I don't see a way around it, given that many permission checks rely on running methods or reading attributes on the model class.

I think the answer is quite simply that these other methods become an implementation detail, to hide the complexity. Also note that I'm referring to the create permission specifically in this comment. If indeed it's a Model-level permission then I think it should be possible to make such a check into a class method, or at least if an instance is required as an implementation detail we can hide that from the dev.

Re: your other points, they are true, but if it's possible to program defensively then I think we should not prematurely discard the idea. For example if there's only a 1% chance that someone will write an error but we write 10000 LOCs, then we'll still expect to see a lot of errors. Can we eliminate or diminish the possibility of an error? If so, how much effort will it take and is it worthwhile? "Worthwhile" is perhaps a matter of opinion -- so what do you think?